### PR TITLE
Making gadgets work on other distributions than Flatcar Edge

### PR DESCRIPTION
**Pull request by alban**
Some gadgets like execsnoop currently only work on Flatcar Edge. This is because Inspektor Gadget relies on:
- cgroup-v2 enabled in containers
- Linux >= 4.18 (for [`bpf_get_current_cgroup_id`](https://github.com/iovisor/bcc/blob/master/docs/kernel-versions.md))
- runc recompiled with [additional static OCI hooks](https://github.com/kinvolk/runc/tree/alban/static-hooks)

This PR provides an alternative way to make it work without cgroup-v2, without Linux >= 4.18 and without a patched runc.
- It uses BCC option `--mntnsmap` to filter pods based on its mount namespace instead of its cgroup-v2 (https://github.com/iovisor/bcc/pull/2834).
- It uses a ldpreload program installed on the Kubernetes nodes via `/etc/ld.so.preload` to enable OCI PreStart & PostStop hooks. 

Tested successfully with execsnoop on Flatcar Stable.

Depends on:
- [x] https://github.com/iovisor/bcc/pull/2834

### CODE REVIEW

File: cmd/kubectl-gadget/deploy.go [hunk 1]
```diff
@@ -34,6 +35,11 @@ func init() {
 		"traceloop", "",
 		true,
 		"enable the traceloop gadget")
+	deployCmd.PersistentFlags().StringVarP(
+		&runcHooks,
+		"runc-hooks", "",
```
Comment (mauriciovasquezbernal) [inline cmd/kubectl-gadget/deploy.go:40]: What about calling it `runc-hooks-mode`?

File: cmd/kubectl-gadget/deploy.go [hunk 2]
```diff
@@ -34,6 +35,11 @@ func init() {
...
+		"auto",
+		"how to attach runc hooks (auto, ldpreload)")
```
Comment (mauriciovasquezbernal) [inline cmd/kubectl-gadget/deploy.go:42]: I think `auto` could be confusing. `auto` could be interpreted like "you don't have to set it, we'll figure this out for you", but that's not the case currently. What about using "patched" for the case we are using edge with the patched runc version?

File: gadget-container/Makefile [hunk 3]
```diff
@@ -31,6 +31,12 @@ networkpolicyadvisor:
 networkpolicyadvisor/push: networkpolicyadvisor
 	for POD in `kubectl get pod -n kube-system -l k8s-app=gadget -o=jsonpath='{.items[*].metadata.name}'` ; do kubectl cp ./bin/networkpolicyadvisor -n kube-system $$POD:/bin/ ; done
 
+.PHONY: runchookslib
```
Comment (mauriciovasquezbernal) [inline gadget-container/Makefile:34]: nit: does it really need to be a PHONY target?

File: gadget-container/entrypoint.sh [hunk 4]
```diff
@@ -39,20 +39,33 @@ ARGS=k8s
...
+  cp /opt/runchooks/add-hooks.jq /host/opt/runchooks/
+  touch /host/etc/ld.so.preload
+  if grep -q ^/opt/runchooks/runchooks.so$ /host/etc/ld.so.preload > /dev/null ; then
+    echo "runchooks.so already setup in /etc/ld.so.preload"
+  else
+    echo "/opt/runchooks/runchooks.so" >> /host/etc/ld.so.preload
   fi
+fi
+
+if [ "$FLATCAR_EDGE" = 1 ] || [ "$INSPEKTOR_GADGET_OPTION_RUNC_HOOKS" = "ldpreload" ] ; then
```
Comment (mauriciovasquezbernal) [inline gadget-container/entrypoint.sh:67]: I don't understand this line. Shouldn't it compare if `RUNC_HOOKS` is `auto`? Why to install `runc-hook-prestart.sh runc-hook-poststop.sh` if they are not needed in the ldpreload case?

File: gadget-container/runchooks/runchooks.c [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	if (!create_found)
+		return;
+
+	// ignore runc programs without bundle directory
+	if (bundledir == NULL)
+		return;
+
+	char cmd[1024] = {0,};
+
+	ret = snprintf(cmd, sizeof(cmd), "cd %s && cp config.json config.json.orig && cat config.json.orig | jq -r \". * `cat /opt/runchooks/add-hooks.jq`\" > config.json", bundledir);
```
Comment (mauriciovasquezbernal) [inline gadget-container/runchooks/runchooks.c:60]: For the reference https://github.com/kinvolk/inspektor-gadget/pull/41#discussion_r417412917

@alban what do you think about using a C json library to do that work?

File: gadget-container/Makefile (hunk 3 above)
Comment (alban) [reply gadget-container/Makefile:34]: I guess not, it could be named `bin/runchooks.so`

File: cmd/kubectl-gadget/deploy.go (hunk 2 above)
Comment (mauriciovasquezbernal) [reply cmd/kubectl-gadget/deploy.go:42]: Good point, if we choose patched and the user runs in a non Flatcar Edge system, it'll not work. I'm wondering if we can introduce an "auto" mode really quickly, i.e, it'll use "patched" if detects Flatcar Edge, otherwise will switch back to "ldpreload".

File: gadget-container/entrypoint.sh (hunk 4 above)
Comment (mauriciovasquezbernal) [reply gadget-container/entrypoint.sh:67]: Ah got it, I thought add-hooks.jq was invoking `ocihookgadget` directly.

File: gadget-container/runchooks/runchooks.c (hunk 5 above)
Comment (mauriciovasquezbernal) [reply gadget-container/runchooks/runchooks.c:60]: Right, in that case we could create an independent executable that uses the json library to do the work and then invoke from here with `system()`. I have never used any C json library, from the ones linked in https://www.json.org/json-en.html I saw there are many that are composed by a single source and header file, so we could take and compile the code directly. 

Given that it's a hack, maybe the only thing we could do now is to ignore any error if jq is not present on the system. Or do you think it's a good idea to make this mechanism more robust and get rid of the jq dependency?

File: gadget-container/entrypoint.sh (hunk 4 above)
Comment (alban) [reply gadget-container/entrypoint.sh:67]: I did that initially, but then `ocihookgadget` could return errors (e.g. if the gadgettracermanager is no longer running because the gadget pod has been stopped) and that was breaking all new pods.
Now, the scripts like `gadget-container/ocihookgadget/runc-hook-prestart.sh` have this protection:
```
pidof gadgettracermanager > /dev/null || exit 0
```
So new pods don't crash if the gadget pod is no longer running.

Alternatively, we could get the grpc client to avoid returning any errors if the grpc server is not running.

**Comment by alban** [comment]
I rebased the PR on master and fixed the conflict

**Comment by alban** [comment]
I rebased the branch and removed the "WIP" since https://github.com/iovisor/bcc/pull/2834 is merged.

File: gadget-container/entrypoint.sh [hunk 6]
```diff
@@ -39,33 +39,60 @@ ARGS=k8s
...
   if grep -q '^GROUP=edge$' /host/etc/flatcar/update.conf > /dev/null ; then
+    echo "Flatcar Edge detected."
     FLATCAR_EDGE=1
+
+    CGROUP_V1_PATH=$(cat /proc/1/cgroup |grep ^1:|cut -d: -f3)
+    CGROUP_V2_PATH=$(cat /proc/1/cgroup |grep ^0:|cut -d: -f3)
+    if [ $CGROUP_V1_PATH != $CGROUP_V2_PATH ] ; then
+      echo "cgroup-v2 is not correctly enabled on Kubernetes pods" >&2
+      exit 1
+    fi
```
Comment (mauriciovasquezbernal) [inline gadget-container/entrypoint.sh:50]: I'm wondering what is the purpose of this test, and if this should be moved outside the Flatcar Edge case. Also, would exit 1 force the container to be created again?, what about the case where the user wants to run a gadget that doesn't depend on this support?

**Comment by alban** [comment]
@mauriciovasquezbernal I think we can merge this PR, if you're also ok with it.

About the documentation changes: we will need to add the technical requirements for running IG and list platforms where it has been tested. But this can be done in a follow-up PR, and I think it is fine to remove obsolete information.
